modules/log/file_test.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package log
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFileLoggerFails(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFileLogger")
+	assert.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	prefix := "TestPrefix "
+	level := INFO
+	flags := LstdFlags | LUTC | Lfuncname
+	//filename := filepath.Join(tmpDir, "test.log")
+
+	fileLogger := NewFileLogger()
+	//realFileLogger, ok := fileLogger.(*FileLogger)
+	//assert.Equal(t, true, ok)
+
+	// Fail if there is bad json
+	err = fileLogger.Init("{")
+	assert.Error(t, err)
+
+	// Fail if there is no filename
+	err = fileLogger.Init(fmt.Sprintf("{\"prefix\":\"%s\",\"level\":\"%s\",\"flags\":%d,\"filename\":\"%s\"}", prefix, level.String(), flags, ""))
+	assert.Error(t, err)
+
+	// Fail if the file isn't a filename
+	err = fileLogger.Init(fmt.Sprintf("{\"prefix\":\"%s\",\"level\":\"%s\",\"flags\":%d,\"filename\":\"%s\"}", prefix, level.String(), flags, tmpDir))
+	assert.Error(t, err)
+
+}
+
+func TestFileLogger(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFileLogger")
+	assert.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	prefix := "TestPrefix "
+	level := INFO
+	flags := LstdFlags | LUTC | Lfuncname
+	filename := filepath.Join(tmpDir, "test.log")
+
+	fileLogger := NewFileLogger()
+	realFileLogger, ok := fileLogger.(*FileLogger)
+	assert.Equal(t, true, ok)
+
+	location, _ := time.LoadLocation("EST")
+
+	date := time.Date(2019, time.January, 13, 22, 3, 30, 15, location)
+
+	dateString := date.UTC().Format("2006/01/02 15:04:05")
+
+	event := Event{
+		level:    INFO,
+		msg:      "TEST MSG",
+		caller:   "CALLER",
+		filename: "FULL/FILENAME",
+		line:     1,
+		time:     date,
+	}
+
+	expected := fmt.Sprintf("%s%s %s:%d:%s [%c] %s\n", prefix, dateString, event.filename, event.line, event.caller, strings.ToUpper(event.level.String())[0], event.msg)
+
+	fileLogger.Init(fmt.Sprintf("{\"prefix\":\"%s\",\"level\":\"%s\",\"flags\":%d,\"filename\":\"%s\",\"maxsize\":%d,\"compress\":false}", prefix, level.String(), flags, filename, len(expected)*2))
```

Comment:
Since we run our tests in a Linux environment it may not be a huge deal, but on Windows machines this fails to init because the `filename` has backslashes and JSON gets mad. Error is `invalid character 'U' in string escape code` because the filepath starts with `C:\Users\...`  

![log_windows](https://user-images.githubusercontent.com/42128690/55189226-53282080-516b-11e9-99d5-452a89c216bb.png)
  
That causes the next tests to fail because it can't initialize properly.

Reply:
If I change it to `filepath.ToSlash(filename)` the tests pass.

Reply:
Done. I should have just used a `map[string]interface{}` and marshalled it like I do in settings but this felt safe enough for a test.

Happy to do it with the map if thought necessary.